go-controller/pkg/ovn/egressgw.go

```diff
@@ -555,61 +555,63 @@ func (oc *Controller) addGWRoutesForPod(gateways []*gatewayInfo, podIfAddrs []*n
 // there are no gateways
-func deletePerPodGRSNAT(nbClient libovsdbclient.Client, node string, podIPNets []*net.IPNet) error {
-	gr := util.GetGatewayRouterFromNode(node)
-	nats := make([]*nbdb.NAT, 0, len(podIPNets))
-	var nat *nbdb.NAT
+func buildPerPodGRSNAT(nodeName string, watchFactory *factory.WatchFactory, podIPNets []*net.IPNet) ([]*nbdb.NAT, error) {
 	var err error
-	for _, podIPNet := range podIPNets {
-		podIP := podIPNet.IP.String()
-		mask := GetIPFullMask(podIP)
-		_, fullMaskPodNet, err := net.ParseCIDR(podIP + mask)
-		if err != nil {
-			klog.Errorf("Invalid IP: %s and mask: %s combination, error: %v", podIP, mask, err)
-			continue
-		}
-		nat = libovsdbops.BuildRouterSNAT(nil, fullMaskPodNet, "", nil)
-		nats = append(nats, nat)
-	}
-	err = libovsdbops.DeleteNatsFromRouter(nbClient, gr, nats...)
-	if err != nil {
-		return fmt.Errorf("failed to delete SNAT rule for pod on gateway router %s, "+
-			"error: %v", gr, err)
-	}
-	return nil
-}
-
-func addPerPodGRSNAT(nbClient libovsdbclient.Client, watchFactory *factory.WatchFactory, pod *kapi.Pod, podIfAddrs []*net.IPNet) error {
-	nodeName := pod.Spec.NodeName
 	node, err := watchFactory.GetNode(nodeName)
 	if err != nil {
-		return fmt.Errorf("failed to get node %s: %v", nodeName, err)
+		return nil, fmt.Errorf("failed to get node %s: %v", nodeName, err)
 	}
 	l3GWConfig, err := util.ParseNodeL3GatewayAnnotation(node)
 	if err != nil {
-		return fmt.Errorf("unable to parse node L3 gw annotation: %v", err)
+		return nil, fmt.Errorf("unable to parse node L3 gw annotation: %v", err)
 	}
-	nats := make([]*nbdb.NAT, 0, len(l3GWConfig.IPAddresses)*len(podIfAddrs))
+	nats := make([]*nbdb.NAT, 0, len(l3GWConfig.IPAddresses)*len(podIPNets))
 	var nat *nbdb.NAT
-	gr := types.GWRouterPrefix + nodeName
 	for _, gwIPNet := range l3GWConfig.IPAddresses {
 		gwIP := gwIPNet.IP.String()
-		for _, podIPNet := range podIfAddrs {
+		for _, podIPNet := range podIPNets {
 			podIP := podIPNet.IP.String()
 			if utilnet.IsIPv6String(gwIP) != utilnet.IsIPv6String(podIP) {
 				continue
 			}
 			mask := GetIPFullMask(podIP)
 			_, fullMaskPodNet, err := net.ParseCIDR(podIP + mask)
 			if err != nil {
-				return fmt.Errorf("invalid IP: %s and mask: %s combination, error: %v", podIP, mask, err)
+				return nil, fmt.Errorf("invalid IP: %s and mask: %s combination, error: %v", podIP, mask, err)
 			}
 			nat = libovsdbops.BuildRouterSNAT(&gwIPNet.IP, fullMaskPodNet, "", nil)
 			nats = append(nats, nat)
 		}
 	}
+	return nats, nil
+}
+
+// deletePerPodGRSNAT removes per pod SNAT rules that are applied to the GR where the pod resides if
+// there are no gateways
+func deletePerPodGRSNAT(nbClient libovsdbclient.Client, nodeName string, watchFactory *factory.WatchFactory, podIPNets []*net.IPNet) error {
+	gr := types.GWRouterPrefix + nodeName
+	nats, err := buildPerPodGRSNAT(nodeName, watchFactory, podIPNets)
```

Comment:
@tssurya is there any way we could use externalIDs on the NATs to figure out what to delete? eg, if the node's L3GW annotation changes between `addPerPodGRSNAT()` and `deletePerPodGRSNAT()` would that fail to build the same NAT objects as were added, and thus leave some stale ones around?

Reply:
hmm yeah you do make a good point, but how do we know which `externalIDs` should be deleted and which shouldn't be? L3GW should be the source of truth, but if that annotation does change you are right in saying we'll be left with stale entries between a podAdd and podDel.
Workaround would be to go through each of the snat's against this podIP but then we'd need to check if the externalID is an egressIP or nodeIP or staleIP or something? 

OH I have a better idea, how about,, I split this function into `deletePerPodGRSNATTowardsNodeIP` & `deletePerPodGRSNAT`. The former will be used only from the egressIP code when disableSNATs is set while when a pod deletion takes place, we'll keep the older format where we nuke all SNAT's against a pod.

Reply:
@dcbw : ^ wdyt?

Reply:
but I think it won't fix the problem fully wrt to egressIPs I guess... I probably need to pass the egressIPs that are serving a podIP as an externalID and then go through the SNAT's and delete anything that's not in that list, but that's too complex because we might add/del wrong egressIPs and we'd need to add logic to track podIPs being served by egressIPs at a given time?

Reply:
oh wait actually when the L3GW annotation changes we are already handling the part where we update the SNATs, so we should be fine:
1) https://github.com/ovn-org/ovn-kubernetes/blob/cd1a816843332c8bbabe0ca2c9d029fb400d9151/go-controller/pkg/ovn/ovn.go#L1190
2) https://github.com/ovn-org/ovn-kubernetes/blob/cd1a816843332c8bbabe0ca2c9d029fb400d9151/go-controller/pkg/ovn/ovn.go#L1063
3) https://github.com/ovn-org/ovn-kubernetes/blob/cd1a816843332c8bbabe0ca2c9d029fb400d9151/go-controller/pkg/ovn/master.go#L658
4) https://github.com/ovn-org/ovn-kubernetes/blob/cd1a816843332c8bbabe0ca2c9d029fb400d9151/go-controller/pkg/ovn/gateway_init.go#L416-L425

Reply:
hmm actually maybe we are not doing what we should be, if the gatewayIP does change we should be deleting the perPod SNAT's as well in addition to just removing the whole `logicalSubnet`, seems like we are not matching on the externalIDs here as well..